src/composition.rs

```diff
@@ -0,0 +1,649 @@
+                            } else {
+                                false
+                            }
+                        });
+                        return false;
+                    }
+                }
+            }
+        }
+        true
+    }
+}
+
+type CompositionResult = Result<(), ()>;
+
+/// Composition acts as the public entry point into the composition functionality.
+struct Composition {
+    file: ast::File,
+}
+
+impl ToString for Composition {
+    fn to_string(&self) -> String {
+        flux::formatter::convert_to_string(&self.file)
+            .expect("Unable to convert composition file to string.")
+    }
+}
+
+impl Composition {
+    #[allow(dead_code)]
+    fn new(file: ast::File) -> Self {
+        Self { file }
+    }
+
+    /// Initialize an ast::File for use in composition.
+    ///
+    /// This must be called before any other composition can be made, as it'll set up the
+    /// statement that will be managed by composition.
+    #[allow(dead_code)]
+    fn initialize(
+        &mut self,
+        bucket: &str,
+        measurement: Option<&str>,
+    ) -> CompositionResult {
+        let mut visitor =
+            CompositionStatementFinderVisitor::default();
+        flux::ast::walk::walk(
+            &mut visitor,
+            flux::ast::walk::Node::File(&self.file),
+        );
+        if let Some(expr_statement) = visitor.statement {
+            self.file.body = self
+                .file
+                .body
+                .iter()
+                .filter(|statement| match statement {
+                    ast::Statement::Expr(expression) => {
+                        expr_statement != *expression.as_ref()
+                    }
+                    _ => true,
+                })
+                .cloned()
+                .collect();
+        }
+
+        let statement = match measurement {
+            None => {
+                pipe!(
+                    ast::Expression::PipeExpr(Box::new(pipe!(
+                        ast::Expression::Call(Box::new(from!(
+                            bucket.to_string()
+                        ))),
+                        range!()
+                    ))),
+                    yield_!()
+                )
+            }
+            Some(measurement) => {
+                pipe!(
+                    ast::Expression::PipeExpr(Box::new(pipe!(
+                        ast::Expression::PipeExpr(Box::new(pipe!(
+                            ast::Expression::Call(Box::new(from!(
+                                bucket.to_string()
+                            ))),
+                            range!()
+                        ),)),
+                        filter!(
+                            "_measurement".into(),
+                            measurement.into()
+                        )
+                    ))),
+                    yield_!()
+                )
+            }
+        };
+
+        self.file.body.insert(
+            0,
```

Comment:
We always insert at the front of the body.

So even if we have a previous flux script (and eventually, a previous composition)...we'll still initialize at the front. Instead of append at the back. Is this what we want?

Reply:
I am not even _considering_ multiple compositions right now, and we always want to be at the top of the file, yes.

Reply:
ok